### PR TITLE
increase coverage and minor bug fixes

### PR DESCRIPTION
**Pull request by ranjithkumar007**
While writing test cases , i came across several bugs as listed below and these are fixed now.
- [acsc()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-7adabd929d7b40deac32003312d6f910R281) and [asec()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-7adabd929d7b40deac32003312d6f910R286) in real_double.cpp
- [Beta::rewrite_as_gamma()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-ccb9531ddf99a9263a8ca4275bdf438eR2758) in functions.cpp
- [acsch()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-748ec5462b257fcab8881d2d458be573R904), [acoth()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-748ec5462b257fcab8881d2d458be573R953), [asech()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-748ec5462b257fcab8881d2d458be573R973) and [2 bugs in acosh()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-748ec5462b257fcab8881d2d458be573R914) in real_mpfr.cpp
- removed [extra check](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-0c21d17fbbe529072b91e762a5c9760dL421) in parser.cpp
- [cot()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-f9e408919c4b9935fa9a23ee4661686fR679),[csc()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-f9e408919c4b9935fa9a23ee4661686fR699),[sec()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-f9e408919c4b9935fa9a23ee4661686fR689) ,[coth()](https://github.com/symengine/symengine/compare/master...ranjithkumar007:TargetCov100%25?expand=1#diff-f9e408919c4b9935fa9a23ee4661686fR813) in complex_mpc.cpp

Apart from the above bug fixes, vector of tuples is used at many places to facilitate testing and remove code redundancy.

Doubts & further TO-DO's : 
- [x]  In [log](https://github.com/symengine/symengine/blob/master/symengine/functions.cpp#L1222) ```is_canonical()``` , why argument used is of type ```const Basic``` instead of ```const RCP<const Basic>``` ?
- [x]  tests for ```is_canonical``` in hyperbolic functions and increase coverage of ```logic.cpp```
- [x] Fix the bug in comparing two objects of type [ComplexMPC](https://github.com/symengine/symengine/blob/master/symengine/complex_mpc.cpp#L43). the correct way of comparing them is to decompose output of mpc_cmp function into x and y and then check them as mentioned in [docs](http://www.multiprecision.org/index.php?prog=mpc&page=html#index-mpc_005fcmp).

### CODE REVIEW

**Comment by isuruf** [comment]
Thanks for doing this. I'll have to go through all of them to make sure they are correct. Where did you get the test values from?

Btw, can you fix the test failures with piranha?

**Comment by ranjithkumar007** [comment]
I used SymPy as benchmark for tests..
reg piranha error , the error message shown in travis is not giving me any hint on how to resolve it.

File: symengine/tests/basic/test_functions.cpp [hunk 1]
```diff
@@ -3197,31 +3727,84 @@ TEST_CASE("MPFR and MPC: functions", "[functions]")
...
+                            549306144334055_z),
+            std::make_tuple(acoth(real_mpfr(b4)), 490414626505863_z,
+                            490414626505864_z),
+            std::make_tuple(asech(real_mpfr(b6)), 1566799236972411_z,
+                            1566799236972412_z),
+            std::make_tuple(acsch(real_mpfr(b4)), 440191235352683_z,
+                            440191235352684_z),
+
+            std::make_tuple(log(real_mpfr(b4)), 788457360364270,
+                            788457360364271),
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: Can you try adding `_z` to the integers and try compiling locally with piranha?

File: symengine/complex_mpc.cpp [hunk 2]
```diff
@@ -41,9 +41,13 @@ int ComplexMPC::compare(const Basic &o) const
     const ComplexMPC &s = down_cast<const ComplexMPC &>(o);
     if (get_prec() == s.get_prec()) {
         int cmp = mpc_cmp(this->i.get_mpc_t(), s.i.get_mpc_t());
-        if (cmp == 0)
+        int x = MPC_INEX_RE(cmp), y = MPC_INEX_IM(cmp);
```
Comment (isuruf) [inline symengine/complex_mpc.cpp]: Why this change?

File: symengine/functions.cpp [hunk 3]
```diff
@@ -1216,33 +1216,33 @@ RCP<const Basic> ACsc::create(const RCP<const Basic> &arg) const
 Log::Log(const RCP<const Basic> &arg) : OneArgFunction(arg)
 {
     SYMENGINE_ASSIGN_TYPEID()
-    SYMENGINE_ASSERT(is_canonical(*arg))
+    SYMENGINE_ASSERT(is_canonical(arg))
 }
 
-bool Log::is_canonical(const Basic &arg) const
+bool Log::is_canonical(const RCP<const Basic> &arg) const
```
Comment (isuruf) [inline symengine/functions.cpp]: Why this change?

Comment (ranjithkumar007) [inline symengine/functions.cpp]: All other function's `is_canonical()` takes `const RCP<const Basic> &` as argument type.To maintain consistency , i changed this. Is there any particular reason why only `Log`  has `const Basic &` as argument type ?

File: symengine/complex_mpc.cpp (hunk 2 above)
Comment (ranjithkumar007) [inline symengine/complex_mpc.cpp]: presently in master 
```
    mpc_class c(100);
    mpc_pow_ui(c.get_mpc_t(), a.get_mpc_t(), 2, MPFR_RNDN);
    RCP<const Number> r1 = complex_mpc(std::move(c));

    mpc_class ee(100);
    mpc_set_ui_ui(ee.get_mpc_t(), 20, 14, MPFR_RNDN);
    RCP<const Number> r2 = complex_mpc(std::move(ee));

    std::cout<<r1>compare(*r2) << " , "<<r2->compare(*r1)<<"\n";
    
```

Output is `1 , 1`. # Atleast one should have been `-1`.
[MPC Docs](http://www.multiprecision.org/index.php?prog=mpc&page=html#index-mpc_005fcmp) mentions that the return value could be decomposed into `x` and `y` . and after decomposition , i tried to make compare of mpc close to the way we compare in [complex_double](https://github.com/symengine/symengine/blob/master/symengine/complex_double.cpp#L43-L53)

**Comment by ranjithkumar007** [comment]
@isuruf  , Any idea why this [job](https://travis-ci.org/symengine/symengine/jobs/213379463) is failing ?

File: symengine/tests/basic/test_functions.cpp [hunk 4]
```diff
@@ -2211,6 +2581,27 @@ TEST_CASE("Acosh: functions", "[functions]")
...
+    r1 = acosh(complex_double(std::complex<double>(1, 2)));
+    r2 = acosh(real_double(-1.0));
+    REQUIRE(is_a<ComplexDouble>(*r1));
+    REQUIRE(is_a<ComplexDouble>(*r2));
+    REQUIRE(std::abs(std::abs(down_cast<const ComplexDouble &>(*r1).i)
+                     - 1.90908861124732)
+            < 1e-12);
+    REQUIRE(std::abs(std::abs(down_cast<const ComplexDouble &>(*r2).i)
+                     - 3.14159265358979)
+            < 1e-12);
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: This test fails because of errors in `acosh` in the libm implementation. You can increase the tolerance level to `1e-10`

Comment (ranjithkumar007) [inline symengine/tests/basic/test_functions.cpp]: error log says 
```
0.0000000111 < 0.0
```
So i think i need to increase it atleast to 1e-7 for the test case to run ?

**Comment by ranjithkumar007** [comment]
@isuruf , PR is ready for a review.

**Comment by isuruf** [comment]
Thanks. looks good to me. I'm going to go through this again to make sure the tests have correct values. @srajangarg, can you have a look as well to make sure the values are correct>

**Comment by ranjithkumar007** [comment]
yes, We need to make sure that all these values are correct .. I will do another round of verification locally.

**Comment by srajangarg** [comment]
I checked most of the values and they are correct. Thanks for the changes.

**Comment by isuruf** [comment]
@ranjithkumar007, thanks for improving the test-suite and fixing the bugs. @srajangarg, thanks for reviewing

**Comment by ranjithkumar007** [comment]
thanks for the help @isuruf  @srajangarg , Now master's coverage is at 83% :smile:
